### ISSUE TITLE
SC2319 - confused warning message when checking $? inside a function

### ISSUE THREAD

**Issue by mgsmgit** [koalaman/shellcheck#2571]
Hi all!!

Conducted a test on the site https://www.shellcheck.net
I report that this problem is repeated inside a function:

**Here's a snippet or screenshot that shows the problem:**

```
#!/bin/bash
print_codout() {
	RETURN=$?
	printf '%s\n' "$RETURN"
}
[ -f foo ] || print_codout
```

**Here's what shellcheck currently says:**
```
$ shellcheck myscript

Line 3
RETURN=$?
       ^-- SC2319 (warning): This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten.
```

_Originally posted by @mymegasam in https://github.com/koalaman/shellcheck/issues/2544#issuecomment-1242734669_

**Comment by ThanAngell** [comment]
I concur that this appears to be a incorrect warning. I received the same warning in a function being called from an exit trap.

`cleanup()
{
  rc=$?
  # do cleanup
  return $rc
}

trap 'cleanup;' exit`

running locally with version: 0.9.0

**Comment by joe733** [comment]
Happens here too:

```bash
#!/bin/bash

if [[ $exit_code -ne 0 ]]; then
    :
else
    exit_code=$?    
fi
```

```sh
Line 25:
    exit_code=$?    
              ^-- [SC2319](https://www.shellcheck.net/wiki/SC2319) (warning): This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten.
```

**Comment by Hi-Angel** [comment]
@mgsmgit was it fixed? I don't seem to see references from commits to this issue…

**Comment by mgsmgit** [comment]
> @mgsmgit was it fixed? I don't seem to see references from commits to this issue…
Oh, it's a mistake, I accidentally closed it, I'm sorry.

**Comment by Alexander-Shukaev** [comment]
```sh
    [ "${#}" -gt 0 ] || return "${?}"
```
```text
                                ^--^ SC2319 (warning): This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten.
```

Why?

**Comment by eaibmz** [comment]
```
a=$1
b=$2

[ "$a" = "$b" ]
result=$?           <--------  SC2319 (warning): This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten.

echo "$result"

result=$(ls)

echo "$result"
```
